### PR TITLE
Remove extra dim in cpu mode at scatter

### PR DESCRIPTION
**Pull request by innerlee**
## Motivation

Currently the cpu version of scatter will add an extra dim for tensors, which causes problems such as in https://github.com/open-mmlab/mmocr/pull/327

## Modification

Remove the extra dim

## BC-breaking (Optional)

YES

Downstream repos might be affected. Please check mmdet, mmaction2 and mmpose before merge. Please check their tests and also demos in cpu mode.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmcv/parallel/data_parallel.py [hunk 1]
```diff
@@ -18,13 +18,25 @@ class MMDataParallel(DataParallel):
...
-            Defaults to None when GPU is not available.
+            Use None when GPU is not available. Note this is different
+            from pytorch's DataParallel where None means use all gpu devices.
         output_device (str | int): Device ID for output. Defaults to None.
         dim (int): Dimension used to scatter the data. Defaults to 0.
     """
 
     def __init__(self, *args, dim=0, **kwargs):
-        super(MMDataParallel, self).__init__(*args, dim=dim, **kwargs)
+        device_ids = kwargs.get('device_ids', None)
```
Comment (zhouzaida) [inline mmcv/parallel/data_parallel.py:28]:
```python
        device_ids = kwargs.get('device_ids', None)
        if device_ids is None:
            device_ids = []
        if device_ids == []:
```
-> 
```python
        device_ids = kwargs.get('device_ids',  [])
        if len(device_ids) == 0:
```

Comment (innerlee) [reply mmcv/parallel/data_parallel.py:28]: device_ids could be None (although not here)

**Comment by hhaAndroid** [comment]
@innerlee I tested mmdet and it will report an error for the following reasons：

```shell
class Scatter:

    @staticmethod
    def forward(target_gpus, input):
        input_device = get_input_device(input)
        streams = None
        if input_device == -1 and target_gpus != [-1]:
            # Perform CPU to GPU copies in a background stream
            streams = [_get_stream(device) for device in target_gpus]
        # input shape=(1,3,100,100)
        # outputs shape=(1,3,100,100)
        outputs = scatter(input, target_gpus, streams)
        # Synchronize with the copy stream
        if streams is not None:
            synchronize_stream(outputs, target_gpus, streams)

        return tuple(outputs) # (3,100,100),
```
Because the batch dimension is missing, subsequent inferences will report an error.

**Comment by innerlee** [comment]
@hhaAndroid do you know where the batch dim was added in gpu code?

**Comment by hhaAndroid** [comment]
> @hhaAndroid do you know where the batch dim was added in gpu code?

The collate function output of dataloader includes the batch dimension.

**Comment by innerlee** [comment]
Okay so we should make the behavior of collate consistent across cpu/gpu, right?

**Comment by hhaAndroid** [comment]
> Okay so we should make the behavior of collate consistent across cpu/gpu, right?

No, your understanding is wrong. The `collate` does not consider CPU or GPU, the output always includes batch.

**Comment by innerlee** [comment]
> The collate function output of dataloader includes the batch dimension.

Where is the extra dim added for gpu tensors?

**Comment by hhaAndroid** [comment]
> > The collate function output of dataloader includes the batch dimension.
> 
> Where is the extra dim added for gpu tensors?

Sorry, I am late in replying, I will provide you with a detailed description document.

**Comment by innerlee** [comment]
Cool, if you got extra time, please take a look at this bug https://github.com/open-mmlab/mmocr/pull/327. Checkout that branch and run this line https://github.com/open-mmlab/mmocr/pull/327/files#diff-2f44c053cce5ce1b42ec4cd7d50aca219b338cefe5aa8831813d6f4020fca740R14

**Comment by HAOCHENYE** [comment]
Solved by #1621

**Pull request closed without merging**